Adding classical limit in anharmonic free energy

OrbitalC opened the pull request:
Adding a --classical option in anharmonic_free energy to ease the comparison with thermodynamic integration.
I don't use the full high temperature simplification (setting $n(\omega) + 1 \approx \frac{k_BT}{\hbar\omega}$ because of the $n(\omega)>>1$ but instead I simply redefine $n(\omega)\approx \frac{k_BT}{\hbar\omega}$ and use the same formula for the anharmonic contributions. It's simpler in the code and should give the same results when the high-temperature limit is actually applicable.)

I also corrected the formula for the third order contribution to the free energy which had a wrong occupation number (a $n(\omega_1)$ instead of $n(\omega_2)$.

flokno commented:
Thanks!

Short discussion: I think we should try to be consistent with other parts of the code, both in terms of occupation and flags. E.g., `canonical_configuration` offers the flag `--quantum` to switch to BE statistics.

I would therefore suggest to do it the same way here. Also in terms of occupation.

How much work do you think this would be?

OrbitalC commented:
I think it's a good idea to be consistent, though I'm wondering if it wouldn't make more sense for the default behavior to be quantum ? And actually the same would apply for the `canonical_configuration` binary, where I think the most common usage is with the `--quantum` flag. But this would change a lot of script/framework.
In any case, I can do the change quickly.

I also realize that the units are not consistent in the output of this binary where some quantities are in eV/atoms and others in eV/f.u. I need to change this before this can be merge.

flokno commented:
I agree that little speaks against making `--quantum` the default. It's a bit a philosophy question whether you come from MD or sampling.

However, simply out of pragmatism, I'd go with adding `--quantum` here in `anharmonic_free_energy` similar to the other binaries. As long as the behavior is consistent and predictive, it's not much of a deal to carry around the `--quantum` flag

flokno commented:
@OrbitalC ready to :shipit: ?

OrbitalC commented:
I need to redo the outfile.anharmonic_free_energy file to make it easy to parse with python/numpy.
Also, I think it would be a good idea to add some other thermodynamic properties, like entropy or specific heat.
I have the formulas needed, and I could use this branch to add it up all at the same time.

flokno commented:
> I need to redo the outfile.anharmonic_free_energy file to make it easy to parse with python/numpy. Also, I think it would be a good idea to add some other thermodynamic properties, like entropy or specific heat. I have the formulas needed, and I could use this branch to add it up all at the same time.

Ok. I think the easiest is sth like this:
https://github.com/tdep-developers/tdep/blob/bb6e8f994a87ecc870444d0435305aa36269e7d0/src/extract_forceconstants/helper.f90#L12-L25

which gives an output file similar to

```
# FC3 stress (symmetrized), unit: GPa
# stress (Voigt xx, yy, zz, yz, xz, xy)
  -3.76247980E-01 -3.76247980E-01 -3.76247980E-01  0.00000000E+00  0.00000000E+00  0.00000000E+00
# std. dev. (Voigt xx, yy, zz, yz, xz, xy)
   4.01558047E-02  4.01558047E-02  4.01558047E-02  0.00000000E+00  0.00000000E+00  0.00000000E+00
# std. err. (Voigt xx, yy, zz, yz, xz, xy)
   1.02459616E-03  1.02459616E-03  1.02459616E-03  0.00000000E+00  0.00000000E+00  0.00000000E+00
   ```

A better format would be JSON but we don't have any routines for that in TDEP so I'd recommend against that.

Also possibe would be csv with e.g.

```
temperature,free_energy,...
300,value,...
```

which parses directly into sth and could be extended later for a temperature range, though Olle is probably against that

OrbitalC commented:
So I completed things on this merge request.

- Added options for thirdorder and fourthorder : Instead of looking for the infile.forceconstant_XXX, anharmonic corrections are triggered by those variables. (fourthorder automatically set thirdorder to `.true.`
- Added a variable stochastic : When sampling is done with MD, the second order cumulant comes with a + sign. When it's done with phonons, it comes with a - sign. By default, the second order cumulant is with a plus sign (so MD) and the `--stochastic` variable change it to a minus sign (phonon sampling)
- Added also a third order cumulant to the stdout to allow for an estimation of the accuracy.
-  Redid the output, so that it can be parsed with numpy. It looks like
```
 # Free energy at   800.00 K, unit : eV/atom
 # Lowest order (1st order cumulant, 2nd order cumulant)
  -1377.81055  -1377.81719 
 # Third order anharmonic corrections (1st order cumulant, 2nd order cumulant)
  -1377.81434  -1377.81590 
 # Fourth order anharmonic corrections (1st order cumulant, 2nd order cumulant)
  -1377.81462  -1377.81611 
```

And the end of the stdout looks like
```
 Lowest order approximation to the free energy: (eV/atom)
 Calculated as <U - U_second - U_polar> + F_phonon
 F (eV/atom) =   -1377.8105485616131     
 F_phonon (eV/atom) = -0.35689667124469099     
 Second order cumulant (meV/atom):   6.6424749757930668     
 Third order cumulant (meV/atom):   1.2531739886456705     
 
 Free energy with third order anharmonic corrections: (eV/atom)
 Calculated as <U - U_second - U_polar - U_third> + F_phonon + F_3
 F (eV/atom) =   -1377.8143382684168     
 F_3 (eV/atom) =   -3.7514069112498313E-003
 Second order cumulant (meV/atom):   1.5568494693469372     
 Third order cumulant (meV/atom):  0.34047059948225028     
 
 Free energy with fourth order anharmonic corrections: (eV/atom)
 Calculated as <U - U_second - U_polar - U_third - U_fourth> + F_phonon + F_3 + F_4
 F (eV/atom) =   -1377.8146236850741     
 F_3 (eV/atom) =   -3.7514069112498313E-003
 F_4 (eV/atom) =    4.9657808433868601E-004
 Second order cumulant (meV/atom):   1.4850499330410505     
 Third order cumulant (meV/atom):  -1.9351580220259099E-002
```

I also corrected the test to account for the change in the output

flokno commented:
Thanks!

2 comments:

- [ ] should we dump more than 5 digits in the free energy in eV/atom? I guess people will be interest rather in the meVs, then it's only 2 more digits. Or maybe dump meV/atom right away with 3 more digits or so? Not sure how many will be significant anyways
- [ ] Can you align the output a bit more, vertically, in the stdout? Maybe

```
 F (eV/atom) =                       -1377.8105485616131     
 F_phonon (eV/atom) =                    -0.35689667124469099     
 Second order cumulant (meV/atom):        6.6424749757930668     
 Third order cumulant (meV/atom):         1.2531739886456705     
```
just to make me happy :)

flokno commented:
@OrbitalC anything left to do?

OrbitalC commented:
With my last commit, the stdout looks like

``` Lowest order approximation to the free energy (eV/atom):
 Calculated as <U - U_second - U_polar> + F_phonon
                     F =        -1377.8249535
              F_phonon =           -0.3737916
 Second order cumulant =            0.0097199
  Third order cumulant =           -0.0086496
 
 Free energy with third order anharmonic corrections (eV/atom):
 Calculated as <U - U_second - U_polar - U_third> + F_phonon + F_3
                     F =        -1377.8173401
              F_phonon =           -0.3737916
                   F_3 =           -0.0033313
 Second order cumulant =            0.0024267
  Third order cumulant =            0.0015396
 
 Free energy with fourth order anharmonic corrections (eV/atom):
 Calculated as <U - U_second - U_polar - U_third - U_fourth> + F_phonon + F_3 + F_4
                    F  =        -1377.8173589
              F_phonon =           -0.3737916
                   F_3 =           -0.0033313
                   F_4 =            0.0076443
 Second order cumulant =            0.0023070
  Third order cumulant =           -0.0007182
```

If that's ok with you, we are good to go

flokno reviewed (commented):
It looks good, last thing while you're add it: Can you format with 

```
fprettify -i 4 -l 500 --disable-indent-mod *f90
```

?

OrbitalC commented:
Done !

flokno commented:
Thanks!

flokno reviewed (approved):
:shipit:
